python/paddle/python_wrapper_demo/block.py

```diff
@@ -0,0 +1,135 @@
+import paddle.v2.framework.core as core
+
+__all__ = [
+    'Block',
+    'g_block',
+    'block',
+    'eval',
+]
+
+
+class Block(object):
+    '''
+    Block is the concept of code block, which has a sequence of local Variables
+    and Operators.
+    '''
+
+    def __init__(self):
+        '''
+        namespace: str
+        '''
+        self.cmds = []
+
+    def append(self, cmd):
+        '''
+        cmd: Block or Op or Layer
+        '''
+        self.cmds.append(cmd)
+
+    def execute(self):
+        '''
+        Execute this block, this will run all the operators and update the coresponding
+        output variables.
+        '''
+        self._build_nn()
+        self.net.run()
+
+    def _build_nn(self):
+        self.net = core.Net.create()
+        ops = self.__extract_op_from_block(self.cmds)
+        for op in ops:
+            self.net.append_op(op)
+        self.net.complete_add_op(True)
+
+    def __extract_op_from_block(self, cmds):
+        ops = []
+        for cmd in cmds:
+            if type(cmd) is Block:
+                child_ops = self.__extract_op_from_block([cmd])
+                ops += child_ops
+            else:
+                ops.append(cmd)
+        return ops
+
+
+g_block = Block()
+
+
+#TODO this need to be renamed
+class block_guard(object):
+    '''
+    a wrapper for Block, which automatically change g_block, Namespace.
+
+    usage:
+
+        import paddle as pd
+
+        with pd.block() as block:
+            a = pd.Variable()
+            b = pd.Variable()
+            c = pd.add_two(a, b)
+            block.execute()
+    '''
+    cur_block = g_block
+    last_block = None
+    counter = 0
+
+    def __init__(self, namespace='', block=None, execute_immediately=True):
+        '''
+        namespace: str
+            current block's namespace, if leave default, father's namespace will be used.
+        block: Block
+            current block, if set None, a new Block will be created.
+        execute_immediately: bool
+            if execute_immediately is True, then all the operators of this block will be
+            inserted into father's block immediately, if false, this block is independent
+            from father's block.
+        '''
+        self.namespace = namespace if namespace else block_guard.inc_counter()
+        self.block = block if block else Block()
+        self.execute_immediately = execute_immediately
+
+    def __enter__(self):
+        Namespace.begin(self.namespace)
+        block_guard.last_block = block_guard.cur_block
+        block_guard.cur_block = self.block
+
+        if self.execute_immediately:
+            block_guard.last_block.append(block_guard.cur_block)
+
+    def __exit__(self):
+        Namespace.end()
+        block_guard.cur_block = block_guard.last_block
```

Comment:
Here the current block is dropped in Python. Which object could store that block? Should that block be freed?

Reply:
sub-block will inserted into its father-block, so it will not be freed

```c++
// father-scope [ op1, op2, sub-block, op3, op4]

{// father-scope
   {// sub-scope
   }
}
```